File: runtime/java/cyana-ast-runtime/src/cyana/ast/runtime/ByteBufferedInputStream.java
package cyana.ast.runtime;

import java.io.IOException;
import java.io.InputStream;

/**
 * 类似BufferedInpustream功能的流.
 *
 * @author tian wei jun
 */
public class ByteBufferedInputStream {
  private final int standardBufferCapacity = 1024;
  public int nextReadIndex = 0;
  private int eof = -1;
  private int nextPos = 0;
  private int count = 0;
  private int mark = eof;
  private byte[] buffer;
  private InputStream byteInputStream;

  /**
   * 带参构造.
   *
   * @param byteInputStream 所识别的文本输入流
   */
  public ByteBufferedInputStream(InputStream byteInputStream) {
    this.byteInputStream = byteInputStream;
    init();
  }

  /** 初始化. */
  public void init() {
    buffer = new byte[standardBufferCapacity];
  }

  /**
   * 读取一个字符.
   *
   * @return 所读取的字符，若没有文本了，返回-1
   */
  public int read() {
    int read = eof;
    if (nextPos < count) {
      read = buffer[nextPos++];
      ++nextReadIndex;
    } else {
      if (fill()) {
        read = read();
      }
    }
    return read;
  }

  /**
   * 只有当读完缓冲后才能调用，其他时候不能调用，只能在read方法中调用.
   *
   * @return
   */
  private boolean fill() {
    if (mark == eof) { // 重新填入
      return fillByNoMark();
    } else if (mark == 0) { // 扩容
      return fillByExpansion();
    } else { // mark>0 && mark<=count 移动
      for (int indexOfBuffer = mark; indexOfBuffer < count; indexOfBuffer++) {
        buffer[indexOfBuffer - mark] = buffer[indexOfBuffer];
      }
      int oldCount = count - mark;
      int newCount = oldCount;
      for (int indexOfBuffer = oldCount; indexOfBuffer < buffer.length; indexOfBuffer++) {
        int read = doRead();
        if (eof == read) {
          break;
        }
        buffer[indexOfBuffer] = (byte) read;
        ++newCount;
      }
      this.count = newCount;
      this.nextPos -= mark;
      this.mark = 0;
      return newCount <= count ? false : true;
    }
  }

  private boolean fillByExpansion() {
    int nsz = count + standardBufferCapacity;
    byte[] nBuffer = new byte[nsz];
    System.arraycopy(buffer, 0, nBuffer, 0, count);
    int newCount = count;
    for (int indexOfNewBuffer = count; indexOfNewBuffer < nBuffer.length; indexOfNewBuffer++) {
      int read = doRead();
      if (eof == read) {
        break;
      }
      nBuffer[indexOfNewBuffer] = (byte) read;
      ++newCount;
    }
    this.buffer = nBuffer;
    this.count = newCount;
    return newCount <= count ? false : true;
  }

  private boolean fillByNoMark() {
    int read = doRead();
    if (eof == read) {
      return false;
    }
    mark = eof;
    nextPos = 0;
    count = 0;

    int indexOfFill = 0;
    buffer[indexOfFill++] = (byte) read;
    ++count;

    while (indexOfFill < buffer.length) {
      read = doRead();
      if (eof == read) {
        break;
      }
      buffer[indexOfFill++] = (byte) read;
      ++count;
    }
    return true;
  }

  /**
   * 跳过skipSteps.
   *
   * @param skipSteps 跳过的步数
   */
  public void skip(int skipSteps) {
    for (int skipTimes = 0; skipTimes < skipSteps; skipTimes++) {
      read();
    }
  }

  /** 重置到mark的位置. */
  public void reset() {
    if (mark == eof) {
      return;
    }
    nextReadIndex = nextReadIndex - (nextPos - mark);
    nextPos = mark;
    mark = eof;
  }

  public void mark() {
    this.mark = nextPos;
  }

  private int doRead() {
    int read = eof;
    try {
      read = byteInputStream.read();
    } catch (IOException e) {
      throw new CyanaAstRuntimeException(e);
    }
    return read;
  }

  /** 关闭输入流. */
  public void close() {
    try {
      byteInputStream.close();
    } catch (IOException e) {
      throw new CyanaAstRuntimeException(e);
    }
  }
}

File: runtime/cpp/cyana-ast-runtime-lib/ByteBufferedInputStream.cpp
//
// Created by tian wei jun on 2022/12/2 0002.
//

#include "ByteBufferedInputStream.h"
#include "CyanaAstRuntimeException.h"

int ByteBufferedInputStream::standardBufferCapacity = 256;

ByteBufferedInputStream::ByteBufferedInputStream() : nextReadIndex(0), eof(-1), nextPos(0),
                                                     count(0), markPos(-1), sizeOfBuffer(0),
                                                     buffer(nullptr), byteInputStream(nullptr) {
}

ByteBufferedInputStream::~ByteBufferedInputStream() {
  delete[] buffer;
  buffer = nullptr;

  if (byteInputStream) {
    byteInputStream->close();
    delete byteInputStream;
    byteInputStream = nullptr;
  }
}

int ByteBufferedInputStream::read() {
  int readedByte = eof;
  if (nextPos < count) {
    readedByte = buffer[nextPos++];
    ++nextReadIndex;
  } else {
    if (fill()) {
      readedByte = read();
    }
  }
  return readedByte;
}

/*
 * 只有当读完缓冲后才能调用，其他时候不能调用，只能在read方法中调用.
 */
bool ByteBufferedInputStream::fill() {
  if (markPos == eof) {// 重新填入
    return fillByNoMark();
  } else if (markPos == 0) {// 扩容
    return fillByExpansion();
  } else {// mark>0 && mark<=count 移动
    for (int indexOfBuffer = markPos; indexOfBuffer < count; indexOfBuffer++) {
      buffer[indexOfBuffer - markPos] = buffer[indexOfBuffer];
    }
    int oldCount = count - markPos;
    int newCount = oldCount;
    for (int indexOfBuffer = oldCount; indexOfBuffer < sizeOfBuffer; indexOfBuffer++) {
      int read = doRead();
      if (eof == read) {
        break;
      }
      buffer[indexOfBuffer] = (byte) read;
      ++newCount;
    }
    this->count = newCount;
    this->nextPos -= markPos;
    this->markPos = 0;
    return newCount > oldCount;
  }
}

bool ByteBufferedInputStream::fillByExpansion() {
  int nsz = count + standardBufferCapacity;
  byte *nBuffer = new byte[nsz];
  for (int i = 0; i < count; i++) {
    nBuffer[i] = buffer[i];
  }
  int newCount = count;
  for (int indexOfNewBuffer = count; indexOfNewBuffer < nsz; indexOfNewBuffer++) {
    int read = doRead();
    if (eof == read) {
      break;
    }
    nBuffer[indexOfNewBuffer] = (byte) read;
    ++newCount;
  }
  delete[] this->buffer;
  this->buffer = nBuffer;
  this->count = newCount;
  this->sizeOfBuffer = nsz;
  return newCount > count;
}

bool ByteBufferedInputStream::fillByNoMark() {
  int read = doRead();
  if (eof == read) {
    return false;
  }
  markPos = eof;
  nextPos = 0;
  count = 0;

  int indexOfFill = 0;
  buffer[indexOfFill++] = (byte) read;
  ++count;

  while (indexOfFill < sizeOfBuffer) {
    read = doRead();
    if (eof == read) {
      break;
    }
    buffer[indexOfFill++] = (byte) read;
    ++count;
  }
  return true;
}

void ByteBufferedInputStream::skip(int skipSteps) {
  for (int skipTimes = 0; skipTimes < skipSteps; skipTimes++) {
    read();
  }
}

void ByteBufferedInputStream::reset() {
  if (markPos == eof) {
    return;
  }
  nextReadIndex = nextReadIndex - (nextPos - markPos);
  nextPos = markPos;
  markPos = eof;
}

void ByteBufferedInputStream::mark() {
  this->markPos = nextPos;
}

int ByteBufferedInputStream::doRead() {
  int read = byteInputStream->get();
  if (byteInputStream->fail()) {
    read = -1;
  }
  return read;
}

void ByteBufferedInputStream::init(const std::string *sourceFilePath) {
  nextReadIndex = 0;
  eof = -1;
  nextPos = 0;
  count = 0;
  markPos = -1;
  if (!buffer) {
    buffer = new byte[standardBufferCapacity];
    sizeOfBuffer = standardBufferCapacity;
  }

  if (!byteInputStream) {
    byteInputStream = new std::ifstream();
  } else {
    byteInputStream->close();
    byteInputStream->clear();
  }
  byteInputStream->open(*sourceFilePath, std::ios::in | std::ios::binary);
  if (!byteInputStream->is_open()) {
    throw CyanaAstRuntimeException("open source File error,path:'" + *sourceFilePath + "'");
  }
}